File: desktop/macos-swiftui/Sources/HelloTimeCyber/Models/APIModels.swift
import Foundation

// MARK: - Unified API envelope (mirrors ApiResponse<T> from TypeScript types)
struct APIEnvelope<T: Decodable>: Decodable {
    let success: Bool
    let data: T?
    let message: String?
    let errorCode: String?
}

// MARK: - Core domain models

struct Capsule: Codable, Identifiable, Sendable {
    let code: String
    let title: String
    let content: String?
    let creator: String
    let openAt: Date
    let createdAt: Date
    let opened: Bool?

    var id: String { code }
    var isUnlocked: Bool { opened ?? (openAt <= Date()) }
}

struct CreateCapsuleRequest: Codable, Sendable {
    let title: String
    let content: String
    let creator: String
    let openAt: Date
}

struct AdminLoginRequest: Codable, Sendable {
    let password: String
}

struct AdminToken: Codable, Sendable {
    let token: String
}

struct TechStack: Codable, Sendable {
    let framework: String
    let language: String
    let database: String
}

struct HealthInfo: Codable, Sendable {
    let status: String
    let timestamp: Date
    let techStack: TechStack
}

struct PageData<T: Codable & Sendable>: Codable, Sendable {
    let content: [T]
    let totalElements: Int
    let totalPages: Int
    let number: Int
    let size: Int
}

// MARK: - Empty response sentinel (for DELETE 204)
struct EmptyResponse: Codable, Sendable {}

File: backends/ktor/src/main/kotlin/com/hellotime/model/ApiModels.kt
package com.hellotime.model

import kotlinx.serialization.Serializable

@Serializable
data class ApiResponse<T>(
    val success: Boolean,
    val data: T? = null,
    val message: String? = null,
    val errorCode: String? = null
)

@Serializable
data class TechStack(
    val framework: String,
    val language: String,
    val database: String
)

@Serializable
data class HealthData(
    val status: String,
    val timestamp: String,
    val techStack: TechStack
)

@Serializable
data class CreateCapsuleRequest(
    val title: String,
    val content: String,
    val creator: String,
    val openAt: String
)

@Serializable
data class CapsuleCreatedResponse(
    val code: String,
    val title: String,
    val creator: String,
    val openAt: String,
    val createdAt: String
)

@Serializable
data class CapsuleDetailResponse(
    val code: String,
    val title: String,
    val content: String? = null,
    val creator: String,
    val openAt: String,
    val createdAt: String,
    val opened: Boolean
)

@Serializable
data class AdminLoginRequest(val password: String)

@Serializable
data class AdminTokenResponse(
    val token: String,
    val expiresIn: Long
)

@Serializable
data class AdminCapsuleItem(
    val code: String,
    val title: String,
    val content: String,
    val creator: String,
    val openAt: String,
    val createdAt: String,
    val opened: Boolean
)

@Serializable
data class CapsulePageResponse(
    val content: List<AdminCapsuleItem>,
    val number: Int,
    val size: Int,
    val totalElements: Long,
    val totalPages: Int
)
